cardano-ledger-core/src/Cardano/Ledger/Val.hs

```diff
@@ -100,10 +103,17 @@ instance Val Coin where
   inject = id
   size _ = 1
   modifyCoin f v = f v
-  pointwise p (Coin x) (Coin y) = p x y
+  pointwise = coerce
```

Comment:
This is no faster in this case and is less readable (more magic) so I'd advise against.

Using `coerce` makes sense when you're not just unwrapping one newtype (since GHC already does that by doing zero work) but a bunch of unwrappings (e.g. linear number) that avoids a whole identity traversal.

Reply:
In this case it is not a simple type coerce, because it actually coerces the function `p :: (Integer -> Integer -> Bool)` to `Coin -> Coin -> Bool`. So it is not just individual `x` and `y` unwrappings. Out of curiosity I did check out the core for some simple examples and despite that coercions that were produced looked slightly different they are indeed were the same in the end, so it does look like ghc is pretty smart, which is great!

One way or another less magical is a really good argument. I'll change it back.